src/RegistryCompilerPass.php

```diff
@@ -58,23 +73,35 @@ private function buildRegistry(ContainerBuilder $container, $registryId, array $
 					}
 				}
 
-				$collection->add($id, $options);
+				$collection->add([$id, $optionsResolver->getMetadata($options)], $options);
 			}
 
 		}
 
 		$targetId = $registryId;
 		$methodName = $tagOptions['method'];
+		$useMetadata = null !== $tagOptions['metadata'];
+
 		if ($tagOptions['use_collection']) {
 			$targetId = "{$registryId}.collection";
-			$container->setDefinition($targetId, new Definition(\ArrayObject::class))->setPublic(false);
+
+			$collectionClass = $useMetadata ? MetaDataIterator::class : \ArrayObject::class;
+			$container->setDefinition($targetId, new Definition($collectionClass))->setPublic(false);
```

Comment:
It makes it cleaner this way. The block below is the same whether or not i’m using `use_collection`.

I could instead use `addMethodCall($setter, [iterator_to_array($iteartor)]`, but how would I pass metadata then? Dunno, maybe I should drop `use_collection` altogether?